src/transformation/visitors/binary-expression/compound.ts

```diff
@@ -180,13 +181,86 @@ export function transformCompoundAssignmentStatement(
         return [objAndIndexDeclaration, assignStatement];
     } else {
+        if (isSetterSkippingCompoundAssignmentOperator(operator)) {
+            const luaLhs = context.transformExpression(lhs) as lua.AssignmentLeftHandSideExpression;
+            return transformSetterSkippingCompoundAssignment(context, luaLhs, operator, rhs, node);
+        }
+
         // Simple statements
         // ${left} = ${left} ${replacementOperator} ${right}
         const operatorExpression = transformBinaryOperation(context, left, right, operator, node);
         return transformAssignment(context, lhs, operatorExpression);
     }
 }
+
+/* These setter-skipping operators will not execute the setter if result does not change.
+ * x.y ||= z does NOT call the x.y setter if x.y is already true.
+ * x.y &&= z does NOT call the x.y setter if x.y is already false.
+ * x.y ??= z does NOT call the x.y setter if x.y is already not nullish.
+ */
+type SetterSkippingCompoundAssignmentOperator =
+    | ts.SyntaxKind.AmpersandAmpersandToken
+    | ts.SyntaxKind.BarBarToken
+    | ts.SyntaxKind.QuestionQuestionToken;
+
+function isSetterSkippingCompoundAssignmentOperator(
+    operator: ts.BinaryOperator
+): operator is SetterSkippingCompoundAssignmentOperator {
+    return (
+        operator === ts.SyntaxKind.AmpersandAmpersandToken ||
+        operator === ts.SyntaxKind.BarBarToken ||
+        operator === ts.SyntaxKind.QuestionQuestionToken
+    );
+}
+
+function transformSetterSkippingCompoundAssignment(
+    context: TransformationContext,
+    lhs: lua.AssignmentLeftHandSideExpression,
+    operator: SetterSkippingCompoundAssignmentOperator,
+    rhs: ts.Expression,
+    node: ts.Node
+): lua.Statement[] {
+    if (operator === ts.SyntaxKind.AmpersandAmpersandToken) {
+        return [
+            lua.createIfStatement(
+                lhs,
+                lua.createBlock([lua.createAssignmentStatement(lhs, context.transformExpression(rhs))]),
+                undefined,
+                node
+            ),
+        ];
+    } else if (operator === ts.SyntaxKind.BarBarToken) {
+        return [
+            lua.createIfStatement(
+                lua.createUnaryExpression(lhs, lua.SyntaxKind.NotOperator),
+                lua.createBlock([lua.createAssignmentStatement(lhs, context.transformExpression(rhs))]),
+                undefined,
+                node
+            ),
+        ];
+    } else if (operator === ts.SyntaxKind.QuestionQuestionToken) {
+        return [
+            lua.createIfStatement(
+                lua.createBinaryExpression(lhs, lua.createNilLiteral(), lua.SyntaxKind.EqualityOperator),
+                lua.createBlock([lua.createAssignmentStatement(lhs, context.transformExpression(rhs))]),
+                undefined,
+                node
+            ),
+        ];
+    } else {
+        context.diagnostics.push(unsupportedNodeKind(node, operator));
+        return [];
```

Comment:
Since `operator` is a union, and all it's values are handled above, you can use `assertNever` to statically verify that this branch would never be executed
```suggestion
	assertNever(operator);
```